gdpr_assist/models.py

```diff
@@ -246,33 +282,311 @@ def _cast_class(cls, model, privacy_meta):
+            )
+            indent_level = 0
+
+            def tabify(x):
+                return "\t" * x if x > 0 else ""
+
+            for line in lines:
+                if line["event"] == EventLog.EVENT_RECURSIVE_START:
+                    res += "{tabs}Starting recursive for {model_name} #{target_pk}.\n".format(
+                        tabs=tabify(indent_level),
+                        model_name=line['model_name'],
+                        target_pk=line['target_pk'],
+                    )
+                    indent_level += 1
+
+                elif line["event"] == EventLog.EVENT_RECURSIVE_END:
+                    indent_level -= 1
+                    res += "{tabs}Ending recursive for {model_name} #{target_pk}.\n".format(
+                        tabs=tabify(indent_level),
+                        model_name=line['model_name'],
+                        target_pk=line['target_pk'],
+                    )
+
+                elif line["event"] == EventLog.EVENT_ALREADY_ANONYMISED:
+                    res += "{tabs}{model_name} #{target_pk} already anonymised.\n".format(
+                        tabs=tabify(indent_level),
+                        model_name=line['model_name'],
+                        target_pk=line['target_pk'],
+                    )
+
+                else:
+                    res += "{tabs}{model_name} #{target_pk} flat fields anonymised.\n".format(
+                        tabs=tabify(indent_level),
+                        model_name=line['model_name'],
+                        target_pk=line['target_pk'],
+                    )
+
+            return res
+        else:
+            return "%s is anonymised, but does not have matching logs." % self
+
+    @classmethod
+    def get_anonymisation_tree(cls, prefix="", doprint=False, objs=[]):
+        """
+        Print the result of the nesting defined above for a given model.
+        Useful for sanity and loop checking. Example output:
+
+        Class:
+        |-> outcome_assessments = (OutcomeAssessment [set_field]):
+            |-> score
+            |-> comment
+        |-> class_polls = (ClassPoll [set_field]):
+            |-> poll = (Poll [fk]):
+                |-> title
+                |-> question
+                |-> pollchoice_set = (PollChoice [set_field]):
+                    |-> choice
+                |-> pollsession_set = (PollSession [set_field]):
+                    |-> outcome_assessments = (OutcomeAssessment [set_field]):
+                        |-> score
+                        |-> comment
+                    |-> pollresponse_set = (PollResponse [set_field]):
+                        |-> response
+
+        """
+        ABANDON_AFTER_N_LEVELS = 10
+        BASE_PREFIX = "    "
+        if len(prefix) > len(BASE_PREFIX) * ABANDON_AFTER_N_LEVELS:
+            return "ERROR: shouldn't go %s levels deep, check for loops." % ABANDON_AFTER_N_LEVELS
```

Comment:
@anochens Shouldn't we raise a [`RecursionError`](https://docs.python.org/3/library/exceptions.html#RecursionError) and fail loudly instead?

Reply:
You'll just have to handle it elsewhere. Given that this is a library, it is probably right to raise the error, and handle it application-side. Practically, I like just having a string to not have to deal with it elsewhere. But, you have a good point